apps/rush-lib/src/logic/PackageChangeAnalyzer.ts

```diff
@@ -125,19 +134,59 @@ export class PackageChangeAnalyzer {
     //  });
     // }
 
-    // Determine the current variant from the link JSON.
-    const variant: string | undefined = this._rushConfiguration.currentInstalledVariant;
-
-    // Add the shrinkwrap file to every project's dependencies
-    const shrinkwrapFile: string = path.relative(
-      this._rushConfiguration.rushJsonFolder,
-      this._rushConfiguration.getCommittedShrinkwrapFilename(variant)
-    ).replace(/\\/g, '/');
+    if (
+      this._rushConfiguration.packageManager === 'pnpm' &&
+      this._rushConfiguration.pnpmOptions.incrementalBuildDependencyGranularity === 'project'
+    ) {
+      const projects: RushConfigurationProject[] = [];
+      const projectDependencyManifestPaths: string[] = [];
+
+      for (const project of this._rushConfiguration.projects) {
+        const dependencyManifestFilePath: string = PnpmProjectDependencyManifest.getFilePathForProject(project);
+        const relativeDependencyManifestFilePath: string = path.relative(
+          this._rushConfiguration.rushJsonFolder,
+          dependencyManifestFilePath
+        ).replace(/\\/g, '/');
+
+        if (!FileSystem.exists(dependencyManifestFilePath)) {
+          throw new Error(
+            `A project dependency file (${relativeDependencyManifestFilePath}) is missing. You may need to run ` +
+            '"rush unlink" and "rush link".'
+          );
+        }
+
+        projects.push(project);
+        projectDependencyManifestPaths.push(relativeDependencyManifestFilePath);
+      }
 
-    for (const project of this._rushConfiguration.projects) {
-      const shrinkwrapHash: string | undefined = noProjectHashes[shrinkwrapFile];
-      if (shrinkwrapHash) {
-        projectHashDeps.get(project.packageName)!.files[shrinkwrapFile] = shrinkwrapHash;
+      const hashes: Map<string, string> = gitHashFiles(
+        projectDependencyManifestPaths,
+        this._rushConfiguration.rushJsonFolder
+      );
+      for (let i: number = 0; i < projects.length; i++) {
```

Comment:
Is there any reason this logic can't be merged into the for loop directly above?

Reply:
In the first loop, we collect all of the package dependency manifest files, and then we get all of their hashes, then we populate the package-deps files. It's more efficient to call `gitHashFiles` once with two loops.